### PR TITLE
Scripts cleanup

### PR DESCRIPTION
**Pull request by levmckinney**
This PR extracts common behavior from the training scripts into `ingredients.py` and refactors the training/evaluation loop.

### CODE REVIEW

File: tuned_lens/scripts/ingredients.py [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+def world_size() -> int:
+    """Get the world size from torch.distributed."""
+    return dist.get_world_size() if dist.is_initialized() else 1
+
+
+@dataclass
+class Data:
+    """Configuration for the dataset."""
+
+    name: List[str] = field(default_factory=lambda: ["the_pile", "all"], nargs="*")
```
Comment (norabelrose) [inline tuned_lens/scripts/ingredients.py:50]: nit: say `list[str]` not `List[str]`

File: tuned_lens/scripts/ingredients.py [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+
+    split: str = "validation"
+    """Split of the dataset to use."""
+
+    text_column: str = "text"
+    """Column of the dataset containing text to run the model on."""
+
+    revision: Optional[str] = None
+
+    def load(self, tokenizer: PreTrainedTokenizerBase) -> tuple[Dataset, float]:
```
Comment (norabelrose) [inline tuned_lens/scripts/ingredients.py:62]: goood you used `tuple` not `Tuple` here

File: tuned_lens/scripts/ingredients.py [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+class Model:
+    """Configuration for the model and tokenizer."""
+
+    name: str
+    """Name of model to use in the Huggingface Hub."""
+
+    no_cache: bool = field(action="store_true")
+    """Don't permanently cache the model on disk."""
+
+    random_model: bool = field(action="store_true")
```
Comment (norabelrose) [inline tuned_lens/scripts/ingredients.py:95]: I think the `store_true` thing is not actually needed (like 80% confidence)

File: tuned_lens/scripts/ingredients.py [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+                ),
+            )
+        elif self.cpu_offload:
+            raise ValueError("CPU offload requires FSDP.")
+        else:
+            model.to(local_rank())
+            return model
+
+    def shard_lens(self, lens: TunedLens) -> DDP | TunedLens:
+        """Shard a tuned lens using DistributedDataParallel."""
```
Comment (norabelrose) [inline tuned_lens/scripts/ingredients.py:204]: This isn't actually "sharding" the lens; DDP just copies it to all ranks. We could try applying FSDP to the lens although I haven't done that for reasons I don't 100% remember. I think we could do it but it would require making a new wrapper class for both the model and the tuned lens (which returns like, the normal model output and the lens output at all layers) and then wrap that wrapper with `FullyShardedDataParallel`. But that's bad because it materializes all the logits for all the layers in VRAM at once (not to mention that under the hood FSDP would be copying those logits around a lot between ranks). So yeah I think I chose not to include the lens in FSDP for efficiency/memory reasons

File: tuned_lens/scripts/train_loop.py [hunk 5]
```diff
@@ -16,246 +18,362 @@
...
-
-    dl = DataLoader(
-        data.shuffle(seed=args.seed),  # type: ignore[arg-type]
-        batch_size=args.per_gpu_batch_size,
-    )
-    if args.wandb and local_rank == 0:
+from dataclasses import dataclass
+
+
+class LossChoice(enum.Enum):
```
Comment (norabelrose) [inline tuned_lens/scripts/train_loop.py:32]: Sigh, I have a general policy of using `typing.Literal` instead of `enum.Enum` since `enum.Enum` is quite verbose, but it's ultimately your call.

File: tuned_lens/scripts/train_loop.py [hunk 6]
```diff
@@ -16,246 +18,362 @@
...
+        if local_rank() > 0 or not self.wandb:
+            return
+
         import wandb
 
-        *_, name = args.model_name.split("/")
         wandb.init(
-            config=vars(args),
+            config=vars(self),
             entity="eleutherai",
```
Comment (norabelrose) [inline tuned_lens/scripts/train_loop.py:144]: We do need to make this customizable or something

File: tuned_lens/scripts/train_loop.py [hunk 7]
```diff
@@ -16,246 +18,362 @@
...
+                        raise NotImplementedError
+
+                    # Log the loss *before* LASSO regularization
+                    logging_loss = loss.detach()
+                    maybe_all_reduce(logging_loss)
+                    if local_rank() == 0:
+                        metrics[f"loss/{name}"].append(logging_loss)
+
+                    # Add sparsity regularizer
+                    if self.lasso:
```
Comment (norabelrose) [inline tuned_lens/scripts/train_loop.py:354]: I have literally never used this and only included it because Igor did an experiment on it once so it can probably be removed

**Review by norabelrose — APPROVED** [review]
Overall looks great to me, just a couple nits

File: tuned_lens/scripts/ingredients.py (hunk 3 above)
Comment (levmckinney) [reply tuned_lens/scripts/ingredients.py:95]: No it is totally a thing https://github.com/lebrice/SimpleParsing/blob/1567bd19ef803c7703d76343584cc52feab0c00b/examples/custom_args/custom_args_example.py#L71

File: tuned_lens/scripts/train_loop.py (hunk 6 above)
Comment (levmckinney) [reply tuned_lens/scripts/train_loop.py:144]: Ya this definitly needs to get fixed.

File: tuned_lens/scripts/train_loop.py (hunk 5 above)
Comment (levmckinney) [reply tuned_lens/scripts/train_loop.py:32]: Simple parsing does not support this https://github.com/lebrice/SimpleParsing/issues/2.

File: tuned_lens/scripts/ingredients.py (hunk 3 above)
Comment (alexmlong) [reply tuned_lens/scripts/ingredients.py]: Just to add a little from what I learned doing my refactor, I think you can technically get away with either specifying `store_true` or setting a default value, e.g. `random_model: bool = True`. I think the only difference is that specifying `store_true` means that the flag can't have a value specified for it in the command line, which setting a default value won't prevent. TLDR if you want to reproduce the original behavior of `add_argument` you want to use `action="store_true"`. Here's some test code I ran to show the behavior of the two options:

```
# with `action="store_true"`
> python tuned_lens/scripts/test.py --array
args: Options(array=True)

# with a default value of `False`
(env) ➜  tuned-lens git:(41-refactor-argument-parsers) ✗ python tuned_lens/scripts/test.py --array     
args: Options(array=False)

# trying to specify a value on the command line when using `store_true`:
python tuned_lens/scripts/test.py --array True
usage: test.py [-h] [--array]
test.py: error: unrecognized arguments: True

# when using `store_true` but not specifying the value in the command
> python tuned_lens/scripts/test.py
args: Options(array=False)
```
